src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,836 @@
+                        else 
+                        {
+                            this.DrainLease(partition, CloseReason.Shutdown);
+                            this.RenewLease(partition);
+                            RenewedLease = true;
+                            DrainedLease = true;
+                        }
+                        
+                        try
+                        {
+                            await this.partitionTable.UpdateEntityAsync(partition, etag, (TableUpdateMode)1);
+                            this.LogHelper(partition, false, false, RenewedLease, DrainedLease, ReleasedLease, "");
+                            
+                        }
+                        // Catch exception of failed table update due to ETag expired. 
+                        // Re-read the table immediately to get the latest ETag.
+                        catch (RequestFailedException ex) when (ex.Status == 412)
+                        {
+                            this.settings.Logger.PartitionManagerInfo(
+                                this.storageAccountName,
+                                this.settings.TaskHubName,
+                                this.settings.WorkerId,
+                                partition.RowKey,
+                                $"Failed to update {partition.RowKey} due to concurrency. Re-read the table to get the latest Etag.");
+                            throw;
+                        }
+                        // Eat any other exceptions.
+                        catch (Exception exception)
+                        {
+                            this.settings.Logger.PartitionManagerError(
+                                this.storageAccountName,
+                                this.settings.TaskHubName,
+                                this.settings.WorkerId,
+                                partition.RowKey,
+                                $"Unexpected error occurred in shutdown when updating lease {partition.RowKey} : {exception}"
+                                );
+                        }
+                    }
+
+                }
+                bool releasedAllLeases = (ownershipLeaseCount == 0);
+                return releasedAllLeases;
+            }
+
+           
+        }
+
+        /// <summary>
+        ///The Response class describes the behavior of the ReadandWrite method in the PartitionManager worker class. 
+        ///If the virtual machine is about to drain and release a lease, the method sets the WorkonDrain flag to true. 
+        ///If the VM is going to acquire another lease from other VM, it sets the waitforPartition flag to true. 
+        ///When either of these flags is true, the sleep time of the VM changes to 1 second.
+        /// </summary>
+        class ReadTableReponse
+        {
+            //If set to true, it indicates that the VM is working on release lease. 
+            public bool WorkOnDrain { get; set; } = false;
```

Comment:
I think we need to improve these names since I have a hard time understanding what they mean unless I carefully the code and comments. For this one, I suggest:
```suggestion
            public bool IsDrainingPartition { get; set; } = false;
```

Note that you should start Boolean property names with "Is" when possible. This is a best practice in C# coding.